File: 118_PascalsTriangleII.cpp
// Topic: Pascal's Triangle II
// Description: Return the k-th (0-indexed) row of Pascal's triangle.
// Approach: Build iteratively using O(k) space.

#include <bits/stdc++.h>
using namespace std;

vector<int> getRow(int rowIndex)
{
    vector<int> row;
    for (int i = 0; i <= rowIndex; ++i)
    {
        row.insert(row.begin() + i, 1);
        for (int j = i - 1; j > 0; --j) row[j] += row[j - 1];
    }
    return row;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<int> tests = {0, 1, 4};
    for (int i = 0; i < (int)tests.size(); ++i)
    {
        auto row = getRow(tests[i]);
        cout << "Test " << (i + 1) << " (k=" << tests[i] << "): ";
        for (size_t j = 0; j < row.size(); ++j)
            cout << row[j] << (j + 1 == row.size() ? '\n' : ' ');
    }
    return 0;
}

/*
Time Complexity: O(k^2)
Space Complexity: O(k)
*/

File: 118_PascalsTriangleII.py
# Topic: Pascal's Triangle II
# Description: Return the k-th (0-indexed) row of Pascal's triangle.
# Approach: Build iteratively using O(k) space.

from typing import List


class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        row = [1]
        for i in range(1, rowIndex + 1):
            row.append(1)
            for j in range(i - 1, 0, -1):
                row[j] += row[j - 1]
        return row


if __name__ == "__main__":
    tests = [0, 1, 4]
    for i, k in enumerate(tests):
        print(f"Test {i + 1} (k={k}): {Solution().getRow(k)}")

"""
Time Complexity: O(k^2)
Space Complexity: O(k)
"""
